metadata/db.go

```diff
@@ -0,0 +1,180 @@
+package metadata
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/boltdb/bolt"
+	"github.com/containerd/containerd/gc"
+	"github.com/containerd/containerd/log"
+	"github.com/pkg/errors"
+)
+
+type DB struct {
+	db *bolt.DB
+
+	rlock sync.RWMutex
+	mlock sync.RWMutex
+
+	storeL sync.Mutex
+	ss     map[string]*snapshotter
+	cs     *contentStore
+
+	// TODO: Keep track of stats such as pause time, number of collected objects, errors
+	lastCollection time.Time
+}
+
+func NewDB(db *bolt.DB) *DB {
+	return &DB{
+		db: db,
+		ss: map[string]*snapshotter{},
+	}
+}
+
+func (m *DB) View(fn func(*bolt.Tx) error) error {
+	m.rlock.RLock()
+	defer m.rlock.RUnlock()
+	return m.db.View(fn)
+}
+
+func (m *DB) Update(fn func(*bolt.Tx) error) error {
+	m.mlock.RLock()
+	defer m.mlock.RUnlock()
+	return m.db.Update(fn)
+}
+
+func (m *DB) GarbageCollect(ctx context.Context) error {
+	lt1 := time.Now()
+	m.mlock.Lock()
+	defer func() {
+		m.mlock.Unlock()
+		log.G(ctx).WithField("t", time.Now().Sub(lt1)).Debug("mutations unlocked")
```

Comment:
Use `d` for the name, short for "duration". I think we used "duration` in the past.